nixos/modules/services/networking/smartdns.nix

```diff
@@ -0,0 +1,61 @@
+{ lib, pkgs, config, ... }:
+
+with lib;
+
+let
+  inherit (lib.types) attrsOf coercedTo listOf oneOf str int bool;
+  cfg = config.services.smartdns;
+
+  confFile = pkgs.writeText "smartdns.conf" (with generators;
+    toKeyValue {
+      mkKeyValue = mkKeyValueDefault {
+        mkValueString = v:
+          if isBool v then
+            if v then "yes" else "no"
+          else
+            mkValueStringDefault { } v;
+      } " ";
+      listsAsDuplicateKeys =
+        true; # Allowing duplications because we need to deal with multiple entries with the same key.
+    } cfg.settings);
+in {
+  options.services.smartdns = {
+    enable = mkEnableOption "SmartDNS DNS server";
+
+    bindPort = mkOption {
+      type = types.port;
+      default = 53;
+      description = "DNS listening port number.";
+    };
+
+    settings = mkOption {
+      type = let atom = oneOf [ str int bool ];
+      in attrsOf (coercedTo atom toList (listOf atom));
+      example = literalExample ''
+        settings = {
+          bind = ":5353 -no-rule -group example";
+          cache-size = 4096;
+          server-tls = [ "8.8.8.8:853" " 1.1.1.1:853" ];
+          server-https = "https://cloudflare-dns.com/dns-query -exclude-default-group";
+          prefetch-domain = true;
+          speed-check-mode = "ping,tcp:80";
+        };
+      '';
+      description = ''
+        A set that will be generated into configuration file, see the <link xlink:href="https://github.com/pymumu/smartdns/blob/master/ReadMe_en.md#configuration-parameter">SmartDNS README</link> for details of configuration parameters.
+        You could override the options here like <option>services.smartdns.bindPort</option> by writing <literal>settings.bind = ":5353 -no-rule -group example";</literal>
+      '';
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    systemd.packages = [ pkgs.smartdns ];
+    systemd.services.smartdns.wantedBy = [ "multi-user.target" ];
+    services.smartdns.settings.bind = mkDefault ":${toString cfg.bindPort}";
+    environment.etc."smartdns/smartdns.conf".source = confFile;
+    environment.etc."default/smartdns" = {
+      mode = "0644";
```

Comment:
Is this not the default mode for files?

Reply:
Oh, we are symlinking by default. Why not symlink this as well? Could you add a comment if copy is necessary?

Reply:
You mean the 644 file permission?

Reply:
Well, we are not copying here. We just linked that

Reply:
Symlinks just have the permissions of target on Linux. If you change the mode, it will have to be copied.

Reply:
See also https://github.com/NixOS/nixpkgs/blob/37b590dc453242638f8a04221e906c4a72a1213c/nixos/modules/system/etc/etc.nix#L82-L91

Reply:
It is not required to be copied. I have removed the specification of mode.